{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8d82f70",
   "metadata": {},
   "source": [
    "# Mirror repeat analysis of T2T genome\n",
    "## Table of contents <a name=\"TOC\"></a>\n",
    "1. [Expected number of repeats and AG segments](#Expected_number)\n",
    "    1. [Expected number of AG segments](#Expected_number_segm)\n",
    "    2. [Expected number of purine MR](#Expected_number_repeats)\n",
    "    3. [Number of repeats in AG segments](#Number_repeats_in_segments)\n",
    "    4. [Expected number of repeats in AG segments](#Expected_number_repeats_in_segments)\n",
    "2. [Structure of the mirror repeats](#structure)\n",
    "    1. [Patterns](#patterns)\n",
    "    2. [Types of repeats](#repeat_type)\n",
    "3. [Annotation](#annotation)\n",
    "    1. [RepeatMasker database](#RM)\n",
    "    2. [Genes](#genes)\n",
    "    3. [Introns](#introns)\n",
    "    4. [Enhancers](#enhancers)\n",
    "    5. [Promoters](#promoters)\n",
    "    6. [File for heatmap with contributions of each features](#contribution)\n",
    "    7. [TE](#TE)\n",
    "    8. [ISTAT](#istat)\n",
    "4. [Conservation](#conservation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4be02429",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import random\n",
    "import gzip\n",
    "from natsort import index_natsorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "36967661",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\anaconda3\\lib\\site-packages\\Bio\\pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.Seq import Seq "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5f9c7e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "from Bio import SeqIO \n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f1f2d3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_list_XY = list(range(1,23)) + ['X', 'Y']\n",
    "chr_list_XY = ['chr'+str(i) for i in chrom_list_XY]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1c86d0b",
   "metadata": {},
   "source": [
    "Download T2T genome: T2T-CHM13v2.0 https://www.ncbi.nlm.nih.gov/assembly/GCA_009914755.4#/def"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "51d69a20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, X, Y, "
     ]
    }
   ],
   "source": [
    "CHM13_genome = dict()\n",
    "chr_range = 1\n",
    "chr_list = list(range(chr_range,23))\n",
    "with open('../../../TtoT_genome/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna') as fasta_file:\n",
    "    for sequence in SimpleFastaParser(fasta_file):\n",
    "        chrom = re.split('chromosome |,', sequence[0])[1]\n",
    "        if chrom in [str(chrom) for chrom in chr_list]:\n",
    "            CHM13_genome[int(chrom)] = sequence[1].upper()\n",
    "            print(chrom, end=', ')\n",
    "        elif chrom in ['X', 'Y']:\n",
    "            CHM13_genome[chrom] = sequence[1].upper()\n",
    "            print(chrom, end=', ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "da9890bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#input files\n",
    "MRpur = pd.read_csv('data/MRpurine_HumanGenome.csv', sep='\\t')\n",
    "AGsegm = pd.read_csv('data/AGsegments_HumanGenome.out', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3882b96b",
   "metadata": {},
   "source": [
    "## Expected number of repeats and segments\n",
    "\n",
    "<a name='Expected_number'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ed012e0",
   "metadata": {},
   "source": [
    "### Expected number of AG segments\n",
    "\n",
    "<a name='Expected_number_segm'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6b772c5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_list_XY = list(range(1,23)) + ['X', 'Y']\n",
    "base_counts = dict()\n",
    "for chrom in chrom_list_XY:\n",
    "    base_counts[chrom] = pd.Series([CHM13_genome[chrom].count('A'), CHM13_genome[chrom].count('T'), CHM13_genome[chrom].count('G'), CHM13_genome[chrom].count('C')], index = ['A', 'T', 'G', 'C'])\n",
    "base_counts = pd.concat(base_counts).unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9fcb21f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>T</th>\n",
       "      <th>G</th>\n",
       "      <th>C</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>73600418</td>\n",
       "      <td>71705495</td>\n",
       "      <td>52064401</td>\n",
       "      <td>51017014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>72424400</td>\n",
       "      <td>72614972</td>\n",
       "      <td>48914580</td>\n",
       "      <td>48742800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>61130447</td>\n",
       "      <td>60661558</td>\n",
       "      <td>39576837</td>\n",
       "      <td>39737106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>59694025</td>\n",
       "      <td>60069053</td>\n",
       "      <td>37006775</td>\n",
       "      <td>36805092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>54881762</td>\n",
       "      <td>55211290</td>\n",
       "      <td>35979055</td>\n",
       "      <td>35973332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>51932667</td>\n",
       "      <td>52011787</td>\n",
       "      <td>34148104</td>\n",
       "      <td>34034070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>47481945</td>\n",
       "      <td>47705013</td>\n",
       "      <td>32697146</td>\n",
       "      <td>32683324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>43714844</td>\n",
       "      <td>43699677</td>\n",
       "      <td>29464192</td>\n",
       "      <td>29380618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>44172441</td>\n",
       "      <td>44170480</td>\n",
       "      <td>31213290</td>\n",
       "      <td>31061036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>39354993</td>\n",
       "      <td>39425966</td>\n",
       "      <td>28062985</td>\n",
       "      <td>27914190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>39447258</td>\n",
       "      <td>39530773</td>\n",
       "      <td>28118187</td>\n",
       "      <td>28031551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>39526705</td>\n",
       "      <td>39435295</td>\n",
       "      <td>27117915</td>\n",
       "      <td>27244633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>35164669</td>\n",
       "      <td>34882690</td>\n",
       "      <td>21352354</td>\n",
       "      <td>22166973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>29713014</td>\n",
       "      <td>29904472</td>\n",
       "      <td>20793343</td>\n",
       "      <td>20750663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>28001629</td>\n",
       "      <td>29739684</td>\n",
       "      <td>19941117</td>\n",
       "      <td>22070765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>26134354</td>\n",
       "      <td>28260390</td>\n",
       "      <td>20578163</td>\n",
       "      <td>21357467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>23009049</td>\n",
       "      <td>23103987</td>\n",
       "      <td>19145593</td>\n",
       "      <td>19018268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>24176845</td>\n",
       "      <td>24309092</td>\n",
       "      <td>16165887</td>\n",
       "      <td>15890714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>16032707</td>\n",
       "      <td>16263646</td>\n",
       "      <td>14795755</td>\n",
       "      <td>14615256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>18627634</td>\n",
       "      <td>18636660</td>\n",
       "      <td>14655444</td>\n",
       "      <td>14290517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>13128223</td>\n",
       "      <td>13184375</td>\n",
       "      <td>9323964</td>\n",
       "      <td>9454120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>13953165</td>\n",
       "      <td>13839754</td>\n",
       "      <td>11782803</td>\n",
       "      <td>11749204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X</th>\n",
       "      <td>46615932</td>\n",
       "      <td>46694728</td>\n",
       "      <td>30449071</td>\n",
       "      <td>30499835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>21954563</td>\n",
       "      <td>17929049</td>\n",
       "      <td>13373414</td>\n",
       "      <td>9203003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           A         T         G         C\n",
       "1   73600418  71705495  52064401  51017014\n",
       "2   72424400  72614972  48914580  48742800\n",
       "3   61130447  60661558  39576837  39737106\n",
       "4   59694025  60069053  37006775  36805092\n",
       "5   54881762  55211290  35979055  35973332\n",
       "6   51932667  52011787  34148104  34034070\n",
       "7   47481945  47705013  32697146  32683324\n",
       "8   43714844  43699677  29464192  29380618\n",
       "9   44172441  44170480  31213290  31061036\n",
       "10  39354993  39425966  28062985  27914190\n",
       "11  39447258  39530773  28118187  28031551\n",
       "12  39526705  39435295  27117915  27244633\n",
       "13  35164669  34882690  21352354  22166973\n",
       "14  29713014  29904472  20793343  20750663\n",
       "15  28001629  29739684  19941117  22070765\n",
       "16  26134354  28260390  20578163  21357467\n",
       "17  23009049  23103987  19145593  19018268\n",
       "18  24176845  24309092  16165887  15890714\n",
       "19  16032707  16263646  14795755  14615256\n",
       "20  18627634  18636660  14655444  14290517\n",
       "21  13128223  13184375   9323964   9454120\n",
       "22  13953165  13839754  11782803  11749204\n",
       "X   46615932  46694728  30449071  30499835\n",
       "Y   21954563  17929049  13373414   9203003"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "91db9347",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A    0.296372\n",
       "T    0.296089\n",
       "G    0.204255\n",
       "C    0.203284\n",
       "dtype: float64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_counts.sum(axis=0) / base_counts.sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2ead9f4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Len</th>\n",
       "      <th>n_expected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20</td>\n",
       "      <td>3048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21</td>\n",
       "      <td>1526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>23</td>\n",
       "      <td>382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>24</td>\n",
       "      <td>191</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Len  n_expected\n",
       "0   20        3048\n",
       "1   21        1526\n",
       "2   22         764\n",
       "3   23         382\n",
       "4   24         191"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pA = 0.296372\n",
    "pT = 0.296089\n",
    "pG = 0.204255\n",
    "pC = 0.203284\n",
    "pR = pA + pG\n",
    "genome_len = 3117275501\n",
    "n_expected = []\n",
    "for i in range(20, 4535):\n",
    "    num = (pR**(i))*genome_len\n",
    "    n_expected.append(int(round(num, 0)))\n",
    "Len = list(range(20,4535))\n",
    "d = {'Len': Len, 'n_expected': n_expected}\n",
    "df = pd.DataFrame(data=d)\n",
    "df.to_csv('ExpectedNumber_AGsegments_HumanGenome.csv', sep='\\t', index=False)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e3e5cfa",
   "metadata": {},
   "source": [
    "### Expected number of purine MR\n",
    "\n",
    "<a name='Expected_number_repeats'></a>\n",
    "[Return to Table of Contents](#TOC)\n",
    "\n",
    "- generating chromosomes with nucleotides probabilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a1257ad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_counts = dict()\n",
    "for chrom in chrom_list_XY:\n",
    "    base_counts[chrom] = pd.Series([CHM13_genome[chrom].count('A'), CHM13_genome[chrom].count('T'), CHM13_genome[chrom].count('G'), CHM13_genome[chrom].count('C')], index = ['A', 'T', 'G', 'C'])\n",
    "base_counts = pd.concat(base_counts).unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6824dadd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A    0.296372\n",
       "T    0.296089\n",
       "G    0.204255\n",
       "C    0.203284\n",
       "dtype: float64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_counts.sum(axis=0) / base_counts.sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dbe2a5d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note: using larger integers increases time to generate sequence\n",
    "def DNA(length, A_frac = 296, T_frac = 296, G_frac = 204, C_frac = 204):\n",
    "    return ''.join(random.choice('A'*A_frac+'T'*T_frac+'G'*G_frac+'C'*C_frac) for _ in range(length))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ad739491",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_lengths = pd.Series([len(CHM13_genome[seq]) for seq in CHM13_genome], index = chrom_list_XY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "37454d96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Y  "
     ]
    }
   ],
   "source": [
    "random_genome = {}\n",
    "for chrom in chrom_list_XY:\n",
    "    print('\\r' + str(chrom), end=' ')\n",
    "    random_genome[chrom] = DNA(chrom_lengths[chrom])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "542fd7c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for chrom in chrom_list_XY:\n",
    "    outfile = 'data/outputs/chr' + str(chrom) + '_generated.fasta'\n",
    "    with open(outfile, 'w') as fout:\n",
    "        fout.write('>chr'+str(chrom) + ' generated\\n' + random_genome[chrom] + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a42993d9",
   "metadata": {},
   "source": [
    "### Number of repeats in AG segments\n",
    "\n",
    "<a name='Number_repeats_in_segments'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "633dec58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrX chrY "
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start1</th>\n",
       "      <th>start2</th>\n",
       "      <th>end</th>\n",
       "      <th>lsh</th>\n",
       "      <th>dist</th>\n",
       "      <th>first shoulder</th>\n",
       "      <th>second shoulder</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>start_segm</th>\n",
       "      <th>end_segm</th>\n",
       "      <th>len_segm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73343</td>\n",
       "      <td>73369</td>\n",
       "      <td>73379</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>AGAGAGAGAA</td>\n",
       "      <td>AAGAGAGAGA</td>\n",
       "      <td>AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA</td>\n",
       "      <td>73332</td>\n",
       "      <td>73506</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73376</td>\n",
       "      <td>73401</td>\n",
       "      <td>73411</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>AGAAAGAAAG</td>\n",
       "      <td>GAAAGAAAGA</td>\n",
       "      <td>AGAAAGAAAGGGAGAGAAAGAAAAAGAAAGAAAGA</td>\n",
       "      <td>73332</td>\n",
       "      <td>73506</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73390</td>\n",
       "      <td>73412</td>\n",
       "      <td>73429</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>AGAAAGAAAAAGAAAGA</td>\n",
       "      <td>AGAAAGAAAAAGAAAGA</td>\n",
       "      <td>AGAAAGAAAAAGAAAGAAAGAAAGAAAGAAAAAGAAAGA</td>\n",
       "      <td>73332</td>\n",
       "      <td>73506</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73398</td>\n",
       "      <td>73430</td>\n",
       "      <td>73451</td>\n",
       "      <td>21</td>\n",
       "      <td>11</td>\n",
       "      <td>AAAGAAAGAAAGAAAGAAAGA</td>\n",
       "      <td>AGAAAGAAAGAAAGAAAGAAA</td>\n",
       "      <td>AAAGAAAGAAAGAAAGAAAGAAAAAGAAAGAAAGAAAGAAAGAAAG...</td>\n",
       "      <td>73332</td>\n",
       "      <td>73506</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73438</td>\n",
       "      <td>73452</td>\n",
       "      <td>73465</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>AGAAAGAAAGAAA</td>\n",
       "      <td>AAAGAAAGAAAGA</td>\n",
       "      <td>AGAAAGAAAGAAAGAAAGAAAGAAAGA</td>\n",
       "      <td>73332</td>\n",
       "      <td>73506</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr start1 start2    end lsh dist         first shoulder  \\\n",
       "0  chr1  73343  73369  73379  10   16             AGAGAGAGAA   \n",
       "1  chr1  73376  73401  73411  10   15             AGAAAGAAAG   \n",
       "2  chr1  73390  73412  73429  17    5      AGAAAGAAAAAGAAAGA   \n",
       "3  chr1  73398  73430  73451  21   11  AAAGAAAGAAAGAAAGAAAGA   \n",
       "4  chr1  73438  73452  73465  13    1          AGAAAGAAAGAAA   \n",
       "\n",
       "         second shoulder                                         repeat_seq  \\\n",
       "0             AAGAGAGAGA               AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA   \n",
       "1             GAAAGAAAGA                AGAAAGAAAGGGAGAGAAAGAAAAAGAAAGAAAGA   \n",
       "2      AGAAAGAAAAAGAAAGA            AGAAAGAAAAAGAAAGAAAGAAAGAAAGAAAAAGAAAGA   \n",
       "3  AGAAAGAAAGAAAGAAAGAAA  AAAGAAAGAAAGAAAGAAAGAAAAAGAAAGAAAGAAAGAAAGAAAG...   \n",
       "4          AAAGAAAGAAAGA                        AGAAAGAAAGAAAGAAAGAAAGAAAGA   \n",
       "\n",
       "  start_segm end_segm len_segm  \n",
       "0      73332    73506      174  \n",
       "1      73332    73506      174  \n",
       "2      73332    73506      174  \n",
       "3      73332    73506      174  \n",
       "4      73332    73506      174  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outfile = pd.DataFrame(columns = ['chr', 'start1', 'start2', 'end', 'lsh', 'dist', 'first shoulder', 'second shoulder', 'repeat_seq', 'start_segm', 'end_segm', 'len_segm'])\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "for Chr in chr_list_XY:\n",
    "    print(Chr, end=' ')\n",
    "    MRpur_chr = MRpur[MRpur['chr']==Chr]\n",
    "    MRpur_chr = MRpur_chr.reset_index(drop=True)\n",
    "    AGsegm_chr = AGsegm[AGsegm['chr']==Chr]\n",
    "    AGsegm_chr = AGsegm_chr[AGsegm_chr['len']>=40]\n",
    "    AGsegm_chr = AGsegm_chr.reset_index(drop=True)\n",
    "    \n",
    "    \n",
    "    segm = 0\n",
    "    for rep in range(0, len(MRpur_chr)):\n",
    "        rep_start = MRpur_chr.iloc[rep]['start1']\n",
    "        rep_end = MRpur_chr.iloc[rep]['end']\n",
    "        segm_start = AGsegm_chr.iloc[segm]['start']\n",
    "        segm_end = AGsegm_chr.iloc[segm]['end']\n",
    "        \n",
    "        while (rep_start > segm_end) and (segm < len(AGsegm_chr) - 1):\n",
    "            segm += 1\n",
    "            segm_start = AGsegm_chr.iloc[segm]['start']\n",
    "            segm_end = AGsegm_chr.iloc[segm]['end']\n",
    "            \n",
    "        if segm >= len(AGsegm_chr):\n",
    "            break\n",
    "            \n",
    "        if (rep_start >= segm_start) and (rep_end <= segm_end):\n",
    "            repeat = MRpur_chr.iloc[[rep]]\n",
    "            repeat['start_segm'] = segm_start\n",
    "            repeat['end_segm'] = segm_end\n",
    "            repeat['len_segm'] = segm_end - segm_start\n",
    "            outfile = pd.concat([outfile, repeat], ignore_index = True)\n",
    "        \n",
    "outfile.to_csv('data/outputs/MRpur_inAGsegm.csv', sep='\\t', index=False)\n",
    "outfile.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30f67df2",
   "metadata": {},
   "source": [
    "### Expected number of repeats in AG segments\n",
    "\n",
    "<a name='Expected_number_repeats_in_segments'></a>\n",
    "[Return to Table of Contents](#TOC)\n",
    "- generating AG segments in chromosomes with AG content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "30050fc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y "
     ]
    }
   ],
   "source": [
    "for chrom in chrom_list_XY:\n",
    "    print(chrom, end = ' ')\n",
    "    seq_chr = CHM13_genome[chrom]\n",
    "    AGsegm_chr = AGsegm[AGsegm['chr']=='chr'+str(chrom)]\n",
    "    AGsegm_chr = AGsegm_chr.reset_index(drop=True)\n",
    "    \n",
    "    outfile = \"data/outputs/chr\" + str(chrom) + \"_generatedAGsegm.fasta\"\n",
    "    with open(outfile, 'w') as fout:\n",
    "        fout.write(\">\" + Chr +\" with generated AG segments\\n\")\n",
    "        ch = 0\n",
    "        for row in range(0, len(AGsegm_chr)):\n",
    "            segm_start = AGsegm_chr.iloc[segm]['start']\n",
    "            segm_end = AGsegm_chr.iloc[segm]['end']\n",
    "            fout.write(seq_chr[ch:segm_start])\n",
    "            \n",
    "            segm_seq = seq_chr[segm_start:segm_end]\n",
    "            nA = segm_seq.count('A')\n",
    "            probA = nA/(segm_end-segm_start)\n",
    "            probG = 1 - probA\n",
    "            \n",
    "            fout.write(''.join(list(np.random.choice(['A', 'G'], segm_end-segm_start, p=[probA, probG]))))\n",
    "            \n",
    "            ch = segm_end\n",
    "        \n",
    "        fout.write(seq_chr[ch:])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd4b52fe",
   "metadata": {},
   "source": [
    "## Structure of the mirror repeats\n",
    "\n",
    "<a name='structure'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc697c96",
   "metadata": {},
   "source": [
    "### Patterns in MR\n",
    "\n",
    "<a name='patterns'></a>\n",
    "[Return to Table of Contents](#TOC)\n",
    "- finding patterns in purine MR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e14af56c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start1</th>\n",
       "      <th>start2</th>\n",
       "      <th>end</th>\n",
       "      <th>lsh</th>\n",
       "      <th>dist</th>\n",
       "      <th>first shoulder</th>\n",
       "      <th>second shoulder</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>pattern</th>\n",
       "      <th>complexity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50042</td>\n",
       "      <td>50060</td>\n",
       "      <td>50071</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>GGAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAGG</td>\n",
       "      <td>GGAGGGAGGGAGGCAGAAAGGGAGGGAGG</td>\n",
       "      <td>GGAG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50059</td>\n",
       "      <td>50076</td>\n",
       "      <td>50086</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>AAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAA</td>\n",
       "      <td>AAGGGAGGGAGGCAGAAAGGGAGGGAA</td>\n",
       "      <td>-</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50082</td>\n",
       "      <td>50096</td>\n",
       "      <td>50107</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAGCAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50097</td>\n",
       "      <td>50108</td>\n",
       "      <td>50119</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73343</td>\n",
       "      <td>73369</td>\n",
       "      <td>73379</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>AGAGAGAGAA</td>\n",
       "      <td>AAGAGAGAGA</td>\n",
       "      <td>AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA</td>\n",
       "      <td>AG</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start1  start2    end  lsh  dist first shoulder second shoulder  \\\n",
       "0  chr1   50042   50060  50071   11     7    GGAGGGAGGGA     AGGGAGGGAGG   \n",
       "1  chr1   50059   50076  50086   10     7     AAGGGAGGGA      AGGGAGGGAA   \n",
       "2  chr1   50082   50096  50107   11     3    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "3  chr1   50097   50108  50119   11     0    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "4  chr1   73343   73369  73379   10    16     AGAGAGAGAA      AAGAGAGAGA   \n",
       "\n",
       "                             repeat_seq pattern complexity  \n",
       "0         GGAGGGAGGGAGGCAGAAAGGGAGGGAGG    GGAG          4  \n",
       "1           AAGGGAGGGAGGCAGAAAGGGAGGGAA       -         27  \n",
       "2             GGAAGGAAGGAGCAAGGAAGGAAGG    GGAA          4  \n",
       "3                GGAAGGAAGGAAGGAAGGAAGG    GGAA          4  \n",
       "4  AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA      AG          2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patterns = []\n",
    "complexity = []\n",
    "for i in range(0, len(MRpur)):\n",
    "    \n",
    "    # pattern was not found\n",
    "    flag = False\n",
    "    \n",
    "    seq = MRpur.iloc[i]['repeat_seq']\n",
    "    if ('A' not in seq):\n",
    "        patterns.append('G')\n",
    "        complexity.append('1')\n",
    "        continue\n",
    "    if ('G' not in seq):\n",
    "        patterns.append('A')\n",
    "        complexity.append('1')\n",
    "        continue\n",
    "    subseqs = re.findall(r'([AG]{2,}?)\\1+', seq)\n",
    "    for pattern in subseqs:\n",
    "        if ('G' not in pattern) or ('A' not in pattern):\n",
    "            continue\n",
    "        seq_for_align = pattern * (len(seq)//len(pattern))\n",
    "        score = pairwise2.align.globalms(seq, seq_for_align, 1, 0, -0.1, -0.1, score_only=True)\n",
    "        percent = round((score / len(seq)), 2)\n",
    "        if percent >= 0.85:\n",
    "            # repeat consists of patterns \n",
    "            patterns.append(pattern)\n",
    "            # guess that complexity is length of the pattern\n",
    "            num = len(pattern)\n",
    "            complexity.append(num)\n",
    "            flag = True\n",
    "            break\n",
    "    if flag == False:\n",
    "        patterns.append('-')\n",
    "        num = len(seq)\n",
    "        complexity.append(num)\n",
    "\n",
    "MRpur['pattern'] = patterns\n",
    "MRpur['complexity'] = complexity\n",
    "MRpur.to_csv('MRpur_patterns', sep='\\t', index = False)\n",
    "MRpur.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99308501",
   "metadata": {},
   "source": [
    "### Types of repeats\n",
    "\n",
    "<a name='repeat_type'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6406a508",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start1</th>\n",
       "      <th>start2</th>\n",
       "      <th>end</th>\n",
       "      <th>lsh</th>\n",
       "      <th>dist</th>\n",
       "      <th>first shoulder</th>\n",
       "      <th>second shoulder</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>pattern</th>\n",
       "      <th>complexity</th>\n",
       "      <th>repeat_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50042</td>\n",
       "      <td>50060</td>\n",
       "      <td>50071</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>GGAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAGG</td>\n",
       "      <td>GGAGGGAGGGAGGCAGAAAGGGAGGGAGG</td>\n",
       "      <td>GGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50059</td>\n",
       "      <td>50076</td>\n",
       "      <td>50086</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>AAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAA</td>\n",
       "      <td>AAGGGAGGGAGGCAGAAAGGGAGGGAA</td>\n",
       "      <td>-</td>\n",
       "      <td>27</td>\n",
       "      <td>short_complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50082</td>\n",
       "      <td>50096</td>\n",
       "      <td>50107</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAGCAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50097</td>\n",
       "      <td>50108</td>\n",
       "      <td>50119</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73343</td>\n",
       "      <td>73369</td>\n",
       "      <td>73379</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>AGAGAGAGAA</td>\n",
       "      <td>AAGAGAGAGA</td>\n",
       "      <td>AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA</td>\n",
       "      <td>AG</td>\n",
       "      <td>2</td>\n",
       "      <td>STR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start1  start2    end  lsh  dist first shoulder second shoulder  \\\n",
       "0  chr1   50042   50060  50071   11     7    GGAGGGAGGGA     AGGGAGGGAGG   \n",
       "1  chr1   50059   50076  50086   10     7     AAGGGAGGGA      AGGGAGGGAA   \n",
       "2  chr1   50082   50096  50107   11     3    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "3  chr1   50097   50108  50119   11     0    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "4  chr1   73343   73369  73379   10    16     AGAGAGAGAA      AAGAGAGAGA   \n",
       "\n",
       "                             repeat_seq pattern complexity    repeat_type  \n",
       "0         GGAGGGAGGGAGGCAGAAAGGGAGGGAGG    GGAG          4            STR  \n",
       "1           AAGGGAGGGAGGCAGAAAGGGAGGGAA       -         27  short_complex  \n",
       "2             GGAAGGAAGGAGCAAGGAAGGAAGG    GGAA          4            STR  \n",
       "3                GGAAGGAAGGAAGGAAGGAAGG    GGAA          4            STR  \n",
       "4  AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA      AG          2            STR  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MR_type = []\n",
    "for i in range(0, len(MRpur)):\n",
    "    if MRpur.iloc[i]['lsh'] < 22:\n",
    "        if MRpur.iloc[i]['pattern'] == '-':\n",
    "            MR_type.append('short_complex')\n",
    "        else:\n",
    "            MR_type.append('STR')\n",
    "    else:\n",
    "        if MRpur.iloc[i]['pattern'] == '-':\n",
    "            MR_type.append('long_complex')\n",
    "        else:\n",
    "            if len(MRpur.iloc[i]['pattern']) < 22:\n",
    "                MR_type.append('long_TR')\n",
    "            else:\n",
    "                MR_type.append('long_complex')\n",
    "MRpur['repeat_type'] = MR_type\n",
    "#MRpur.to_csv('MRpur_TypeRepeats.csv', '\\t', index=False)\n",
    "MRpur.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea57bf05",
   "metadata": {},
   "source": [
    "## Annotation\n",
    "\n",
    "<a name='annotation'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c097fc27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df is table with annotation (name columns: #chrom, thickStart, thickEnd, feature - class of element, *Name - name of element)\n",
    "# MR - repeats for annotation (columns: start1 - origin coordinate of the first arm, end - end coordinate of the second arm)\n",
    "# chr_list - list of chromosomes names\n",
    "# ignore_names - if True there are no names of elements (only features f. ex. exon or CDS) if False there are names of element (f. ex. AluSx)\n",
    "def annotation(df, MR, chr_list, ignore_names):\n",
    "    def closest_value(input_list, input_value):\n",
    "        arr = np.asarray(input_list)\n",
    "        i = (np.abs(arr - input_value)).argmin()\n",
    "        return i\n",
    "    \n",
    "    # feature - column name, ind - index of row in df\n",
    "    def write_features(MR, feature, value, ind):\n",
    "        if MR.at[ind, feature] == '-':\n",
    "            MR.at[ind, feature] = value\n",
    "        else:\n",
    "            v0 = MR.at[ind, feature]\n",
    "            if type(v0) == list:\n",
    "                v0 = str(v0)\n",
    "            if type(value) != list:\n",
    "                value = '; ' + value\n",
    "            else:\n",
    "                v0 = v0.strip('][').split(', ')\n",
    "            MR.at[ind, feature] = v0 + value\n",
    "        return MR\n",
    "    \n",
    "    # index row in MR\n",
    "    ind = 0\n",
    "    for c in chr_list:\n",
    "        print(c, end=' ')\n",
    "        df_chr = df[df['#chrom']==c]\n",
    "        MR_chr = MR[MR['chr']==c]\n",
    "        \n",
    "        if len(df_chr) == 0:\n",
    "            continue\n",
    "\n",
    "        repeats_starts = MR_chr['start1'].tolist()\n",
    "        df_starts = df_chr['thickStart'].tolist()\n",
    "        repeats_ends = MR_chr['end'].tolist()\n",
    "        df_ends = df_chr['thickEnd'].tolist()\n",
    "        df_features = df_chr['feature'].tolist()\n",
    "        if ignore_names == False:\n",
    "            df_name = df_chr['Name'].tolist()\n",
    "\n",
    "        for rep in range(0, len(repeats_starts)):\n",
    "            # for considering number of chromosome (miss rows for preceding chromosome)\n",
    "            ind_MR = rep+ind\n",
    "            #search the closest number and write index\n",
    "            if __name__ == \"__main__\" :\n",
    "                num=closest_value(df_starts,repeats_starts[rep])\n",
    "                if (repeats_starts[rep] >= df_ends[num]) | (repeats_ends[rep] <= df_starts[num]):\n",
    "                    pass\n",
    "                elif (repeats_starts[rep] >= df_starts[num]) & (repeats_ends[rep] <= df_ends[num]):\n",
    "                    MR = write_features(MR, 'feature', df_features[num], ind_MR)\n",
    "                    if ignore_names == False:\n",
    "                        MR = write_features(MR, 'name', df_name[num], ind_MR)\n",
    "                    MR = write_features(MR, 'position', 'whole', ind_MR)\n",
    "                    MR = write_features(MR, 'feature_coordinates', [df_starts[num], df_ends[num]], ind_MR)\n",
    "                else:\n",
    "                    rep_lsh = repeats_ends[rep] - repeats_starts[rep]\n",
    "                    if (df_starts[num] < repeats_starts[rep]) & ((df_ends[num] - repeats_starts[rep]) > (rep_lsh // 2)):\n",
    "                        MR = write_features(MR, 'feature', df_features[num], ind_MR)\n",
    "                        if ignore_names == False:\n",
    "                            MR = write_features(MR, 'name', df_name[num], ind_MR)\n",
    "                        MR = write_features(MR, 'position', 'part', ind_MR)\n",
    "                        MR = write_features(MR, 'feature_coordinates', [df_starts[num], df_ends[num]], ind_MR)\n",
    "                    elif (df_starts[num] > repeats_starts[rep]) & ((repeats_ends[rep] - df_starts[num]) > (rep_lsh // 2)):\n",
    "                        MR = write_features(MR, 'feature', df_features[num], ind_MR)\n",
    "                        if ignore_names == False:\n",
    "                            MR = write_features(MR, 'name', df_name[num], ind_MR)\n",
    "                        MR = write_features(MR, 'position', 'part', ind_MR)\n",
    "                        MR = write_features(MR, 'feature_coordinates', [df_starts[num], df_ends[num]], ind_MR)\n",
    "                    else:\n",
    "                        pass\n",
    "        ind += len(repeats_starts)\n",
    "    return MR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb3b7bf4",
   "metadata": {},
   "source": [
    "### RepeatMasker database\n",
    "\n",
    "<a name='RM'></a>\n",
    "[Return to Table of Contents](#TOC)\n",
    "\n",
    "Download https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=1568791217_4daOQLVBcLGDgwRTIvQ0udta7Yd0&clade=mammal&org=Human&db=hs1&hgta_group=genes&hgta_track=hub_3671779_refSeqComposite&hgta_table=0&hgta_regionType=genome&position=chr9%3A145%2C458%2C455-145%2C495%2C201&hgta_outputType=gff&hgta_outFileName=RepeatMasker.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "13ab0e88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>reserved</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>blockStarts</th>\n",
       "      <th>id</th>\n",
       "      <th>description</th>\n",
       "      <th>feature</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4163698</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>2705</td>\n",
       "      <td>(ACCCTA)n#Simple_repeat</td>\n",
       "      <td>29</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>2705</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0,2704,0</td>\n",
       "      <td>-1,1,-1</td>\n",
       "      <td>1</td>\n",
       "      <td>2590 2.9 0.6 1.0 chr1 2 2705 (248384623) + (AC...</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>(ACCCTA)n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4163699</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>7536</td>\n",
       "      <td>L1MC3#LINE/L1</td>\n",
       "      <td>223</td>\n",
       "      <td>+</td>\n",
       "      <td>4663</td>\n",
       "      <td>7533</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>4912,600,518,158,0,1001,108,392,3</td>\n",
       "      <td>-1,4663,-1,5528,-1,6131,-1,7141,-1</td>\n",
       "      <td>5</td>\n",
       "      <td>1304 19.6 8.0 2.0 chr1 4664 5263 (248382065) +...</td>\n",
       "      <td>LINE/L1</td>\n",
       "      <td>L1MC3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4163703</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5267</td>\n",
       "      <td>5850</td>\n",
       "      <td>MER34C_v#LTR/ERV1</td>\n",
       "      <td>147</td>\n",
       "      <td>+</td>\n",
       "      <td>5274</td>\n",
       "      <td>5528</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>6,254,322</td>\n",
       "      <td>-1,7,-1</td>\n",
       "      <td>6</td>\n",
       "      <td>1403 14.7 2.0 0.8 chr1 5275 5528 (248381800) +...</td>\n",
       "      <td>LTR/ERV1</td>\n",
       "      <td>MER34C_v</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4163704</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5685</td>\n",
       "      <td>6131</td>\n",
       "      <td>MSTA1#LTR/ERVL-MaLR</td>\n",
       "      <td>148</td>\n",
       "      <td>+</td>\n",
       "      <td>5686</td>\n",
       "      <td>6131</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0,445,0</td>\n",
       "      <td>-1,1,-1</td>\n",
       "      <td>7</td>\n",
       "      <td>2442 14.8 5.8 1.3 chr1 5687 6131 (248381197) +...</td>\n",
       "      <td>LTR/ERVL-MaLR</td>\n",
       "      <td>MSTA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4163705</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8394</td>\n",
       "      <td>8840</td>\n",
       "      <td>MER4A1#LTR/ERV1</td>\n",
       "      <td>106</td>\n",
       "      <td>+</td>\n",
       "      <td>8398</td>\n",
       "      <td>8840</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3,442,0</td>\n",
       "      <td>-1,4,-1</td>\n",
       "      <td>8</td>\n",
       "      <td>3028 10.6 6.1 0.0 chr1 8399 8840 (248378488) +...</td>\n",
       "      <td>LTR/ERV1</td>\n",
       "      <td>MER4A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #chrom  chromStart  chromEnd                     name  score strand  \\\n",
       "4163698   chr1           0      2705  (ACCCTA)n#Simple_repeat     29      +   \n",
       "4163699   chr1           0      7536            L1MC3#LINE/L1    223      +   \n",
       "4163703   chr1        5267      5850        MER34C_v#LTR/ERV1    147      +   \n",
       "4163704   chr1        5685      6131      MSTA1#LTR/ERVL-MaLR    148      +   \n",
       "4163705   chr1        8394      8840          MER4A1#LTR/ERV1    106      +   \n",
       "\n",
       "         thickStart  thickEnd  reserved  blockCount  \\\n",
       "4163698           1      2705         0           3   \n",
       "4163699        4663      7533         0           9   \n",
       "4163703        5274      5528         0           3   \n",
       "4163704        5686      6131         0           3   \n",
       "4163705        8398      8840         0           3   \n",
       "\n",
       "                                blockSizes  \\\n",
       "4163698                           0,2704,0   \n",
       "4163699  4912,600,518,158,0,1001,108,392,3   \n",
       "4163703                          6,254,322   \n",
       "4163704                            0,445,0   \n",
       "4163705                            3,442,0   \n",
       "\n",
       "                                blockStarts  id  \\\n",
       "4163698                             -1,1,-1   1   \n",
       "4163699  -1,4663,-1,5528,-1,6131,-1,7141,-1   5   \n",
       "4163703                             -1,7,-1   6   \n",
       "4163704                             -1,1,-1   7   \n",
       "4163705                             -1,4,-1   8   \n",
       "\n",
       "                                               description        feature  \\\n",
       "4163698  2590 2.9 0.6 1.0 chr1 2 2705 (248384623) + (AC...  Simple_repeat   \n",
       "4163699  1304 19.6 8.0 2.0 chr1 4664 5263 (248382065) +...        LINE/L1   \n",
       "4163703  1403 14.7 2.0 0.8 chr1 5275 5528 (248381800) +...       LTR/ERV1   \n",
       "4163704  2442 14.8 5.8 1.3 chr1 5687 6131 (248381197) +...  LTR/ERVL-MaLR   \n",
       "4163705  3028 10.6 6.1 0.0 chr1 8399 8840 (248378488) +...       LTR/ERV1   \n",
       "\n",
       "              Name  \n",
       "4163698  (ACCCTA)n  \n",
       "4163699      L1MC3  \n",
       "4163703   MER34C_v  \n",
       "4163704      MSTA1  \n",
       "4163705     MER4A1  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = gzip.open(\"../new/Annotation/RepeatMasker.gtf.gz\", mode='rb')\n",
    "column_names = ['seqname', 'start', 'end', 'name', 'score', 'strand', 'i', 'j', 'f']\n",
    "RM = pd.read_csv(f, sep = \"\\t\")\n",
    "name = RM['name'].tolist()\n",
    "feature = [i.split('#')[1] for i in name]\n",
    "name2 = [i.split('#')[0] for i in name]\n",
    "RM['feature'] = feature\n",
    "RM['Name'] = name2\n",
    "RM = RM[RM['#chrom'] != 'chrM']\n",
    "RM = RM[RM['strand'] == '+']\n",
    "RM = RM.sort_values(by=['#chrom', 'thickStart'])\n",
    "RM = RM.sort_values(\n",
    "   by=\"#chrom\",\n",
    "   key=lambda x: np.argsort(index_natsorted(RM[\"#chrom\"]))\n",
    ")\n",
    "RM.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e9507676",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrX chrY "
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start1</th>\n",
       "      <th>start2</th>\n",
       "      <th>end</th>\n",
       "      <th>lsh</th>\n",
       "      <th>dist</th>\n",
       "      <th>first shoulder</th>\n",
       "      <th>second shoulder</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>pattern</th>\n",
       "      <th>complexity</th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>name</th>\n",
       "      <th>position</th>\n",
       "      <th>feature_coordinates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50042</td>\n",
       "      <td>50060</td>\n",
       "      <td>50071</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>GGAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAGG</td>\n",
       "      <td>GGAGGGAGGGAGGCAGAAAGGGAGGGAGG</td>\n",
       "      <td>GGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50059</td>\n",
       "      <td>50076</td>\n",
       "      <td>50086</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>AAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAA</td>\n",
       "      <td>AAGGGAGGGAGGCAGAAAGGGAGGGAA</td>\n",
       "      <td>-</td>\n",
       "      <td>27</td>\n",
       "      <td>short_complex</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50082</td>\n",
       "      <td>50096</td>\n",
       "      <td>50107</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAGCAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50097</td>\n",
       "      <td>50108</td>\n",
       "      <td>50119</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73343</td>\n",
       "      <td>73369</td>\n",
       "      <td>73379</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>AGAGAGAGAA</td>\n",
       "      <td>AAGAGAGAGA</td>\n",
       "      <td>AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA</td>\n",
       "      <td>AG</td>\n",
       "      <td>2</td>\n",
       "      <td>STR</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>(AGAA)n</td>\n",
       "      <td>part</td>\n",
       "      <td>[73349, 73514]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start1  start2    end  lsh  dist first shoulder second shoulder  \\\n",
       "0  chr1   50042   50060  50071   11     7    GGAGGGAGGGA     AGGGAGGGAGG   \n",
       "1  chr1   50059   50076  50086   10     7     AAGGGAGGGA      AGGGAGGGAA   \n",
       "2  chr1   50082   50096  50107   11     3    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "3  chr1   50097   50108  50119   11     0    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "4  chr1   73343   73369  73379   10    16     AGAGAGAGAA      AAGAGAGAGA   \n",
       "\n",
       "                             repeat_seq pattern complexity    repeat_type  \\\n",
       "0         GGAGGGAGGGAGGCAGAAAGGGAGGGAGG    GGAG          4            STR   \n",
       "1           AAGGGAGGGAGGCAGAAAGGGAGGGAA       -         27  short_complex   \n",
       "2             GGAAGGAAGGAGCAAGGAAGGAAGG    GGAA          4            STR   \n",
       "3                GGAAGGAAGGAAGGAAGGAAGG    GGAA          4            STR   \n",
       "4  AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA      AG          2            STR   \n",
       "\n",
       "          feature     name position feature_coordinates  \n",
       "0  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "1  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "2  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "3               -        -        -                   -  \n",
       "4   Simple_repeat  (AGAA)n     part      [73349, 73514]  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MRpur['feature'] = ['-']*len(MRpur)\n",
    "MRpur['name'] = ['-']*len(MRpur)\n",
    "MRpur['position'] = ['-']*len(MRpur)\n",
    "MRpur['feature_coordinates'] = ['-']*len(MRpur)\n",
    "\n",
    "MRpur_annotated = annotation(RM, MRpur, chr_list_XY, False)\n",
    "MRpur_annotated.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92e559a3",
   "metadata": {},
   "source": [
    "### Genes\n",
    "\n",
    "<a name='genes'></a>\n",
    "[Return to Table of Contents](#TOC)\n",
    "\n",
    "Coordinates of genes in TtoT genome\n",
    "https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=1479044017_GvxaoDm1LcBPZJGO8X8KD7QsPevb&clade=hub_3267197&org=hub_3267197_human&db=hub_3267197_GCA_009914755.4&hgta_group=genes&hgta_track=hub_3267197_refSeqComposite&hgta_table=0&hgta_regionType=genome&position=CP068269.2%3A145%2C458%2C455-145%2C495%2C201&hgta_outputType=gff&hgta_outFileName=RepeatMasker.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "950b9650",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1620915</th>\n",
       "      <td>chr1</td>\n",
       "      <td>205300</td>\n",
       "      <td>205629</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1620917</th>\n",
       "      <td>chr1</td>\n",
       "      <td>205456</td>\n",
       "      <td>205629</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1620916</th>\n",
       "      <td>chr1</td>\n",
       "      <td>208474</td>\n",
       "      <td>213723</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1620918</th>\n",
       "      <td>chr1</td>\n",
       "      <td>210907</td>\n",
       "      <td>213723</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1620919</th>\n",
       "      <td>chr1</td>\n",
       "      <td>246378</td>\n",
       "      <td>247209</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #chrom  thickStart  thickEnd feature\n",
       "1620915   chr1      205300    205629    exon\n",
       "1620917   chr1      205456    205629    exon\n",
       "1620916   chr1      208474    213723    exon\n",
       "1620918   chr1      210907    213723    exon\n",
       "1620919   chr1      246378    247209    exon"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = gzip.open(\"genes_T2T-CHM13v2.0.gtf.gz\", mode='rb')\n",
    "column_names = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "genes = pd.read_csv(f, sep = \"\\t\", names=column_names)\n",
    "\n",
    "# add names of chrs\n",
    "l = list(range(1,23))\n",
    "l.reverse()\n",
    "#without 'chr'\n",
    "chrom_list_XY = ['X'] + l + ['Y']\n",
    "chr_list_XY = ['chr'+str(i) for i in chrom_list_XY]\n",
    "chr_list_XY_GB = []\n",
    "GB_ind = genes['seqname'].unique()\n",
    "for seq in range(0, len(GB_ind)):\n",
    "    df2 = genes[genes['seqname'] == GB_ind[seq]]\n",
    "    chr_list_XY_GB.extend([chr_list_XY[seq]]*len(df2))\n",
    "genes['chr'] = chr_list_XY_GB\n",
    "\n",
    "genes = genes[genes['strand'] == '+']\n",
    "genes = genes[['chr', 'start', 'end', 'feature']]\n",
    "genes = genes.sort_values(by=['chr', 'start'])\n",
    "genes = genes.sort_values(\n",
    "   by=\"chr\",\n",
    "   key=lambda x: np.argsort(index_natsorted(genes[\"chr\"]))\n",
    ")\n",
    "\n",
    "genes = genes.rename(columns={\"chr\": \"#chrom\", \"start\": \"thickStart\", \"end\": \"thickEnd\"})\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "67000254",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrX chrY "
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start1</th>\n",
       "      <th>start2</th>\n",
       "      <th>end</th>\n",
       "      <th>lsh</th>\n",
       "      <th>dist</th>\n",
       "      <th>first shoulder</th>\n",
       "      <th>second shoulder</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>pattern</th>\n",
       "      <th>complexity</th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>name</th>\n",
       "      <th>position</th>\n",
       "      <th>feature_coordinates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50042</td>\n",
       "      <td>50060</td>\n",
       "      <td>50071</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>GGAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAGG</td>\n",
       "      <td>GGAGGGAGGGAGGCAGAAAGGGAGGGAGG</td>\n",
       "      <td>GGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50059</td>\n",
       "      <td>50076</td>\n",
       "      <td>50086</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>AAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAA</td>\n",
       "      <td>AAGGGAGGGAGGCAGAAAGGGAGGGAA</td>\n",
       "      <td>-</td>\n",
       "      <td>27</td>\n",
       "      <td>short_complex</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50082</td>\n",
       "      <td>50096</td>\n",
       "      <td>50107</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAGCAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50097</td>\n",
       "      <td>50108</td>\n",
       "      <td>50119</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73343</td>\n",
       "      <td>73369</td>\n",
       "      <td>73379</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>AGAGAGAGAA</td>\n",
       "      <td>AAGAGAGAGA</td>\n",
       "      <td>AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA</td>\n",
       "      <td>AG</td>\n",
       "      <td>2</td>\n",
       "      <td>STR</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>(AGAA)n</td>\n",
       "      <td>part</td>\n",
       "      <td>[73349, 73514]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start1  start2    end  lsh  dist first shoulder second shoulder  \\\n",
       "0  chr1   50042   50060  50071   11     7    GGAGGGAGGGA     AGGGAGGGAGG   \n",
       "1  chr1   50059   50076  50086   10     7     AAGGGAGGGA      AGGGAGGGAA   \n",
       "2  chr1   50082   50096  50107   11     3    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "3  chr1   50097   50108  50119   11     0    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "4  chr1   73343   73369  73379   10    16     AGAGAGAGAA      AAGAGAGAGA   \n",
       "\n",
       "                             repeat_seq pattern complexity    repeat_type  \\\n",
       "0         GGAGGGAGGGAGGCAGAAAGGGAGGGAGG    GGAG          4            STR   \n",
       "1           AAGGGAGGGAGGCAGAAAGGGAGGGAA       -         27  short_complex   \n",
       "2             GGAAGGAAGGAGCAAGGAAGGAAGG    GGAA          4            STR   \n",
       "3                GGAAGGAAGGAAGGAAGGAAGG    GGAA          4            STR   \n",
       "4  AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA      AG          2            STR   \n",
       "\n",
       "          feature     name position feature_coordinates  \n",
       "0  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "1  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "2  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "3               -        -        -                   -  \n",
       "4   Simple_repeat  (AGAA)n     part      [73349, 73514]  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom_list_XY = list(range(1,23)) + ['X', 'Y']\n",
    "chr_list_XY = ['chr'+str(i) for i in chrom_list_XY]\n",
    "MR_annotated_genes = annotation(genes, MRpur_annotated, chr_list_XY, True)\n",
    "MR_annotated_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93a7f46a",
   "metadata": {},
   "source": [
    "### Introns\n",
    "\n",
    "<a name='introns'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "21971380",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start1</th>\n",
       "      <th>start2</th>\n",
       "      <th>end</th>\n",
       "      <th>lsh</th>\n",
       "      <th>dist</th>\n",
       "      <th>first shoulder</th>\n",
       "      <th>second shoulder</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>pattern</th>\n",
       "      <th>complexity</th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>name</th>\n",
       "      <th>position</th>\n",
       "      <th>feature_coordinates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50042</td>\n",
       "      <td>50060</td>\n",
       "      <td>50071</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>GGAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAGG</td>\n",
       "      <td>GGAGGGAGGGAGGCAGAAAGGGAGGGAGG</td>\n",
       "      <td>GGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50059</td>\n",
       "      <td>50076</td>\n",
       "      <td>50086</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>AAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAA</td>\n",
       "      <td>AAGGGAGGGAGGCAGAAAGGGAGGGAA</td>\n",
       "      <td>-</td>\n",
       "      <td>27</td>\n",
       "      <td>short_complex</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50082</td>\n",
       "      <td>50096</td>\n",
       "      <td>50107</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAGCAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50097</td>\n",
       "      <td>50108</td>\n",
       "      <td>50119</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73343</td>\n",
       "      <td>73369</td>\n",
       "      <td>73379</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>AGAGAGAGAA</td>\n",
       "      <td>AAGAGAGAGA</td>\n",
       "      <td>AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA</td>\n",
       "      <td>AG</td>\n",
       "      <td>2</td>\n",
       "      <td>STR</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>(AGAA)n</td>\n",
       "      <td>part</td>\n",
       "      <td>[73349, 73514]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start1  start2    end  lsh  dist first shoulder second shoulder  \\\n",
       "0  chr1   50042   50060  50071   11     7    GGAGGGAGGGA     AGGGAGGGAGG   \n",
       "1  chr1   50059   50076  50086   10     7     AAGGGAGGGA      AGGGAGGGAA   \n",
       "2  chr1   50082   50096  50107   11     3    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "3  chr1   50097   50108  50119   11     0    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "4  chr1   73343   73369  73379   10    16     AGAGAGAGAA      AAGAGAGAGA   \n",
       "\n",
       "                             repeat_seq pattern complexity    repeat_type  \\\n",
       "0         GGAGGGAGGGAGGCAGAAAGGGAGGGAGG    GGAG          4            STR   \n",
       "1           AAGGGAGGGAGGCAGAAAGGGAGGGAA       -         27  short_complex   \n",
       "2             GGAAGGAAGGAGCAAGGAAGGAAGG    GGAA          4            STR   \n",
       "3                GGAAGGAAGGAAGGAAGGAAGG    GGAA          4            STR   \n",
       "4  AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA      AG          2            STR   \n",
       "\n",
       "          feature     name position feature_coordinates  \n",
       "0  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "1  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "2  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "3               -        -        -                   -  \n",
       "4   Simple_repeat  (AGAA)n     part      [73349, 73514]  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon = genes[genes['feature']=='exon']\n",
    "start_intron = []\n",
    "end_intron = []\n",
    "chr_intron = []\n",
    "for Chr in chr_list_XY:\n",
    "    df2 = exon[exon['#chrom']==Chr]\n",
    "    starts = df2['thickStart'].tolist()\n",
    "    ends = df2['thickEnd'].tolist()\n",
    "    chr_intron.extend([Chr]*(len(starts)-1))\n",
    "    for j in range(1, len(starts)):\n",
    "        start_intron.append(ends[j-1])\n",
    "        end_intron.append(starts[j])\n",
    "d = {'#chrom': chr_intron, 'thickStart': start_intron, 'thickEnd': end_intron}\n",
    "intron = pd.DataFrame(data=d)\n",
    "intron['Len'] = intron['thickEnd']-intron['thickStart']\n",
    "intron = intron[intron[\"Len\"]>0]\n",
    "intron = intron[['#chrom', 'thickStart', 'thickEnd']]\n",
    "intron.to_csv('intron.txt', sep='\\t', index=False)\n",
    "intron['feature'] = ['intron']*len(intron)\n",
    "MR_annotated_RM_genes_introns = annotation(intron, MR_annotated_genes, chr_list_XY, True)\n",
    "MR_annotated_RM_genes_introns.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d49eae3",
   "metadata": {},
   "source": [
    "### Enhancers\n",
    "\n",
    "<a name='enhancers'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6c3e1fe8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start1</th>\n",
       "      <th>start2</th>\n",
       "      <th>end</th>\n",
       "      <th>lsh</th>\n",
       "      <th>dist</th>\n",
       "      <th>first shoulder</th>\n",
       "      <th>second shoulder</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>pattern</th>\n",
       "      <th>complexity</th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>name</th>\n",
       "      <th>position</th>\n",
       "      <th>feature_coordinates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50042</td>\n",
       "      <td>50060</td>\n",
       "      <td>50071</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>GGAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAGG</td>\n",
       "      <td>GGAGGGAGGGAGGCAGAAAGGGAGGGAGG</td>\n",
       "      <td>GGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50059</td>\n",
       "      <td>50076</td>\n",
       "      <td>50086</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>AAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAA</td>\n",
       "      <td>AAGGGAGGGAGGCAGAAAGGGAGGGAA</td>\n",
       "      <td>-</td>\n",
       "      <td>27</td>\n",
       "      <td>short_complex</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50082</td>\n",
       "      <td>50096</td>\n",
       "      <td>50107</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAGCAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50097</td>\n",
       "      <td>50108</td>\n",
       "      <td>50119</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73343</td>\n",
       "      <td>73369</td>\n",
       "      <td>73379</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>AGAGAGAGAA</td>\n",
       "      <td>AAGAGAGAGA</td>\n",
       "      <td>AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA</td>\n",
       "      <td>AG</td>\n",
       "      <td>2</td>\n",
       "      <td>STR</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>(AGAA)n</td>\n",
       "      <td>part</td>\n",
       "      <td>[73349, 73514]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start1  start2    end  lsh  dist first shoulder second shoulder  \\\n",
       "0  chr1   50042   50060  50071   11     7    GGAGGGAGGGA     AGGGAGGGAGG   \n",
       "1  chr1   50059   50076  50086   10     7     AAGGGAGGGA      AGGGAGGGAA   \n",
       "2  chr1   50082   50096  50107   11     3    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "3  chr1   50097   50108  50119   11     0    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "4  chr1   73343   73369  73379   10    16     AGAGAGAGAA      AAGAGAGAGA   \n",
       "\n",
       "                             repeat_seq pattern complexity    repeat_type  \\\n",
       "0         GGAGGGAGGGAGGCAGAAAGGGAGGGAGG    GGAG          4            STR   \n",
       "1           AAGGGAGGGAGGCAGAAAGGGAGGGAA       -         27  short_complex   \n",
       "2             GGAAGGAAGGAGCAAGGAAGGAAGG    GGAA          4            STR   \n",
       "3                GGAAGGAAGGAAGGAAGGAAGG    GGAA          4            STR   \n",
       "4  AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA      AG          2            STR   \n",
       "\n",
       "          feature     name position feature_coordinates  \n",
       "0  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "1  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "2  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "3               -        -        -                   -  \n",
       "4   Simple_repeat  (AGAA)n     part      [73349, 73514]  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_names = ['#chrom', 'thickStart', 'thickEnd']\n",
    "enhancer = pd.read_csv('enhancers_T2T.bed', sep='\\t', names=column_names)\n",
    "enhancer['feature'] = ['enhancer']*len(enhancer)\n",
    "MR_annotated_enhancer = annotation(enhancer, MR_annotated_RM_genes_introns, chr_list_XY, True)\n",
    "MR_annotated_enhancer.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "460af619",
   "metadata": {},
   "source": [
    "### Promoters\n",
    "\n",
    "<a name='promoters'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "103c1a8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrX chrY "
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start1</th>\n",
       "      <th>start2</th>\n",
       "      <th>end</th>\n",
       "      <th>lsh</th>\n",
       "      <th>dist</th>\n",
       "      <th>first shoulder</th>\n",
       "      <th>second shoulder</th>\n",
       "      <th>repeat_seq</th>\n",
       "      <th>pattern</th>\n",
       "      <th>complexity</th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>name</th>\n",
       "      <th>position</th>\n",
       "      <th>feature_coordinates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50042</td>\n",
       "      <td>50060</td>\n",
       "      <td>50071</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>GGAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAGG</td>\n",
       "      <td>GGAGGGAGGGAGGCAGAAAGGGAGGGAGG</td>\n",
       "      <td>GGAG</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50059</td>\n",
       "      <td>50076</td>\n",
       "      <td>50086</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>AAGGGAGGGA</td>\n",
       "      <td>AGGGAGGGAA</td>\n",
       "      <td>AAGGGAGGGAGGCAGAAAGGGAGGGAA</td>\n",
       "      <td>-</td>\n",
       "      <td>27</td>\n",
       "      <td>short_complex</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50082</td>\n",
       "      <td>50096</td>\n",
       "      <td>50107</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAGCAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>G-rich</td>\n",
       "      <td>whole</td>\n",
       "      <td>[50040, 50125]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50097</td>\n",
       "      <td>50108</td>\n",
       "      <td>50119</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>GGAAGGAAGGA</td>\n",
       "      <td>AGGAAGGAAGG</td>\n",
       "      <td>GGAAGGAAGGAAGGAAGGAAGG</td>\n",
       "      <td>GGAA</td>\n",
       "      <td>4</td>\n",
       "      <td>STR</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>73343</td>\n",
       "      <td>73369</td>\n",
       "      <td>73379</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>AGAGAGAGAA</td>\n",
       "      <td>AAGAGAGAGA</td>\n",
       "      <td>AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA</td>\n",
       "      <td>AG</td>\n",
       "      <td>2</td>\n",
       "      <td>STR</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>(AGAA)n</td>\n",
       "      <td>part</td>\n",
       "      <td>[73349, 73514]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start1  start2    end  lsh  dist first shoulder second shoulder  \\\n",
       "0  chr1   50042   50060  50071   11     7    GGAGGGAGGGA     AGGGAGGGAGG   \n",
       "1  chr1   50059   50076  50086   10     7     AAGGGAGGGA      AGGGAGGGAA   \n",
       "2  chr1   50082   50096  50107   11     3    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "3  chr1   50097   50108  50119   11     0    GGAAGGAAGGA     AGGAAGGAAGG   \n",
       "4  chr1   73343   73369  73379   10    16     AGAGAGAGAA      AAGAGAGAGA   \n",
       "\n",
       "                             repeat_seq pattern complexity    repeat_type  \\\n",
       "0         GGAGGGAGGGAGGCAGAAAGGGAGGGAGG    GGAG          4            STR   \n",
       "1           AAGGGAGGGAGGCAGAAAGGGAGGGAA       -         27  short_complex   \n",
       "2             GGAAGGAAGGAGCAAGGAAGGAAGG    GGAA          4            STR   \n",
       "3                GGAAGGAAGGAAGGAAGGAAGG    GGAA          4            STR   \n",
       "4  AGAGAGAGAAAGAGAAAGAAAGAAAGAAGAGAGAGA      AG          2            STR   \n",
       "\n",
       "          feature     name position feature_coordinates  \n",
       "0  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "1  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "2  Low_complexity   G-rich    whole      [50040, 50125]  \n",
       "3               -        -        -                   -  \n",
       "4   Simple_repeat  (AGAA)n     part      [73349, 73514]  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_names = ['#chrom', 'thickStart', 'thickEnd']\n",
    "promoter = pd.read_csv('promoters_T2T.bed', sep='\\t', names=column_names)\n",
    "promoter['feature'] = ['promoter']*len(promoter)\n",
    "MR_annotated_enhancer_promoter = annotation(promoter, MR_annotated_enhancer, chr_list_XY, True)\n",
    "MR_annotated_enhancer_promoter.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "04448cf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "MR_annotated_enhancer_promoter.to_csv('MRpur_annotated.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eed715fc",
   "metadata": {},
   "source": [
    "### File for heatmap with contributions of each features\n",
    "\n",
    "<a name='contribution'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f9f6b634",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_len = 3117275501"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "247ed4ba",
   "metadata": {},
   "source": [
    "calculating number and percent of repeats by type and feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c56dfa0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'STR': 183316, 'short_complex': 63267, 'long_TR': 5905, 'long_complex': 154}\n",
      "{'STR': 66031, 'short_complex': 19695, 'long_TR': 5571, 'long_complex': 102}\n"
     ]
    }
   ],
   "source": [
    "MR_annotated = pd.read_csv('MRpur_annotated.csv', sep='\\t')\n",
    "MRsegm_annotated = pd.read_csv('MRpur_inAGsegm_annotated.csv', sep='\\t')\n",
    "\n",
    "MR_by_repeattype_numb = {}\n",
    "MRsegm_by_repeattype_numb = {}\n",
    "keys1 = list(MR_annotated.repeat_type.unique())\n",
    "keys2 = list(MRsegm_annotated.repeat_type.unique())\n",
    "for key in keys1:\n",
    "    number = len(MR_annotated[MR_annotated['repeat_type'] == key])\n",
    "    MR_by_repeattype_numb[key] = number\n",
    "for key in keys2:\n",
    "    number = len(MRsegm_annotated[MRsegm_annotated['repeat_type'] == key])\n",
    "    MRsegm_by_repeattype_numb[key] = number\n",
    "print(MR_by_repeattype_numb)\n",
    "print(MRsegm_by_repeattype_numb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d468bc3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CDS' 'DNA' 'LINE' 'LTR' 'Low_complexity' 'RC' 'Retroposon' 'SINE'\n",
      " 'Satellite' 'Simple_repeat' 'Unknown' 'enhancer' 'exon' 'intron'\n",
      " 'promoter' 'scRNA' 'srpRNA']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>number</th>\n",
       "      <th>number_pc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STR</td>\n",
       "      <td>CDS</td>\n",
       "      <td>78</td>\n",
       "      <td>0.042549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STR</td>\n",
       "      <td>DNA transposons</td>\n",
       "      <td>225</td>\n",
       "      <td>0.122739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STR</td>\n",
       "      <td>LINE</td>\n",
       "      <td>1865</td>\n",
       "      <td>1.017369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STR</td>\n",
       "      <td>LTR</td>\n",
       "      <td>758</td>\n",
       "      <td>0.413494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>30376</td>\n",
       "      <td>16.570294</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  repeat_type          feature  number  number_pc\n",
       "0         STR              CDS      78   0.042549\n",
       "1         STR  DNA transposons     225   0.122739\n",
       "2         STR             LINE    1865   1.017369\n",
       "3         STR              LTR     758   0.413494\n",
       "4         STR   Low_complexity   30376  16.570294"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = MR_annotated.replace('-', 'Unknown')\n",
    "df['feature'] = df.feature.str.split('; ')\n",
    "df = df.explode(\"feature\")\n",
    "df[['feature1', 'feature2']] = df.feature.str.split(\"/\", expand = True)\n",
    "df = df.reset_index(drop=True)\n",
    "df = df.reset_index()\n",
    "df = df.groupby(['repeat_type', 'feature1'])['index'].count()\n",
    "df = df.to_frame().reset_index()\n",
    "df = df.rename(columns={\"index\": \"number\", 'feature1':'feature'})\n",
    "\n",
    "# add per cent\n",
    "number_per_cent = []\n",
    "for row in range(0, len(df)):\n",
    "    repeat_type = df.iloc[row]['repeat_type']\n",
    "    num = df.iloc[row]['number']\n",
    "    num_repeatType = MR_by_repeattype_numb[repeat_type]\n",
    "    number_per_cent.append((num / num_repeatType)*100)\n",
    "    \n",
    "df['number_pc'] = number_per_cent\n",
    "    \n",
    "print(df.feature.unique())\n",
    "df = df.replace('DNA', 'DNA transposons')\n",
    "df = df.replace('RC', 'RC transposons')\n",
    "df = df.replace('Retroposon', 'SVA')\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "85330ada",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CDS' 'DNA' 'LINE' 'LTR' 'Low_complexity' 'SINE' 'Satellite'\n",
      " 'Simple_repeat' 'Unknown' 'enhancer' 'exon' 'intron' 'promoter' 'srpRNA']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>number</th>\n",
       "      <th>number_pc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STR</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17</td>\n",
       "      <td>0.025745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STR</td>\n",
       "      <td>DNA transposons</td>\n",
       "      <td>23</td>\n",
       "      <td>0.034832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STR</td>\n",
       "      <td>LINE</td>\n",
       "      <td>297</td>\n",
       "      <td>0.449789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STR</td>\n",
       "      <td>LTR</td>\n",
       "      <td>83</td>\n",
       "      <td>0.125699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>22046</td>\n",
       "      <td>33.387348</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  repeat_type          feature  number  number_pc\n",
       "0         STR              CDS      17   0.025745\n",
       "1         STR  DNA transposons      23   0.034832\n",
       "2         STR             LINE     297   0.449789\n",
       "3         STR              LTR      83   0.125699\n",
       "4         STR   Low_complexity   22046  33.387348"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = MRsegm_annotated.replace('-', 'Unknown')\n",
    "df2['feature'] = df2.feature.str.split('; ')\n",
    "df2 = df2.explode(\"feature\")\n",
    "df2[['feature1', 'feature2']] = df2.feature.str.split(\"/\", expand = True)\n",
    "df2 = df2.reset_index(drop=True)\n",
    "df2 = df2.reset_index()\n",
    "df2 = df2.groupby(['repeat_type', 'feature1'])['index'].count()\n",
    "df2 = df2.to_frame().reset_index()\n",
    "df2 = df2.rename(columns={\"index\": \"number\", 'feature1':'feature'})\n",
    "\n",
    "# add per cent\n",
    "number_per_cent = []\n",
    "for row in range(0, len(df2)):\n",
    "    repeat_type = df2.iloc[row]['repeat_type']\n",
    "    num = df2.iloc[row]['number']\n",
    "    num_repeatType = MRsegm_by_repeattype_numb[repeat_type]\n",
    "    number_per_cent.append((num / num_repeatType)*100)\n",
    "    \n",
    "df2['number_pc'] = number_per_cent\n",
    "\n",
    "print(df2.feature.unique())\n",
    "df2 = df2.replace('DNA', 'DNA transposons')\n",
    "df2 = df2.replace('RC', 'RC transposons')\n",
    "df2 = df2.replace('Retroposon', 'SVA')\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3d6e7f69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>number_allMRpur</th>\n",
       "      <th>number_allMRpur_pc</th>\n",
       "      <th>number_MRpur_inAGsegm</th>\n",
       "      <th>number_MRpur_inAGsegm_pc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STR</td>\n",
       "      <td>CDS</td>\n",
       "      <td>78</td>\n",
       "      <td>0.042549</td>\n",
       "      <td>17</td>\n",
       "      <td>0.025745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STR</td>\n",
       "      <td>DNA transposons</td>\n",
       "      <td>225</td>\n",
       "      <td>0.122739</td>\n",
       "      <td>23</td>\n",
       "      <td>0.034832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STR</td>\n",
       "      <td>LINE</td>\n",
       "      <td>1865</td>\n",
       "      <td>1.017369</td>\n",
       "      <td>297</td>\n",
       "      <td>0.449789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STR</td>\n",
       "      <td>LTR</td>\n",
       "      <td>758</td>\n",
       "      <td>0.413494</td>\n",
       "      <td>83</td>\n",
       "      <td>0.125699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>30376</td>\n",
       "      <td>16.570294</td>\n",
       "      <td>22046</td>\n",
       "      <td>33.387348</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  repeat_type          feature  number_allMRpur  number_allMRpur_pc  \\\n",
       "0         STR              CDS               78            0.042549   \n",
       "1         STR  DNA transposons              225            0.122739   \n",
       "2         STR             LINE             1865            1.017369   \n",
       "3         STR              LTR              758            0.413494   \n",
       "4         STR   Low_complexity            30376           16.570294   \n",
       "\n",
       "   number_MRpur_inAGsegm  number_MRpur_inAGsegm_pc  \n",
       "0                     17                  0.025745  \n",
       "1                     23                  0.034832  \n",
       "2                    297                  0.449789  \n",
       "3                     83                  0.125699  \n",
       "4                  22046                 33.387348  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3 = df.rename(columns={\"number\": \"number_allMRpur\", 'number_pc': 'number_allMRpur_pc'})\n",
    "df2 = df2.rename(columns={\"number\": \"number_MRpur_inAGsegm\", 'number_pc': 'number_MRpur_inAGsegm_pc'})\n",
    "df3 = df3.sort_values(by=['repeat_type', 'feature'])\n",
    "df3['number_MRpur_inAGsegm'] = [0]*len(df3)\n",
    "df3['number_MRpur_inAGsegm_pc'] = [0]*len(df3)\n",
    "for i in range(0, len(df3)):\n",
    "    repeat_type = df3.iloc[i]['repeat_type']\n",
    "    feature = df3.iloc[i]['feature']\n",
    "    df4 = df2[df2['repeat_type'] == repeat_type]\n",
    "    if feature in df4.feature.tolist():\n",
    "        row = df4[df4['feature']==feature]\n",
    "        df3.at[i, 'number_MRpur_inAGsegm'] = row.iloc[0]['number_MRpur_inAGsegm']\n",
    "        df3.at[i, 'number_MRpur_inAGsegm_pc'] = row.iloc[0]['number_MRpur_inAGsegm_pc']\n",
    "df3.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c51b18f4",
   "metadata": {},
   "source": [
    "adding contributions of each feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "85c8c808",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['DNA transposons' 'LINE' 'LTR' 'Low_complexity' 'RC transposons'\n",
      " 'Retroposon' 'SINE' 'Satellite' 'Simple_repeat' 'Unknown' 'rRNA' 'scRNA'\n",
      " 'snRNA' 'srpRNA' 'tRNA']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>Len</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNA transposons</td>\n",
       "      <td>62872991</td>\n",
       "      <td>0.020169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LINE</td>\n",
       "      <td>464215444</td>\n",
       "      <td>0.148917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LTR</td>\n",
       "      <td>174930907</td>\n",
       "      <td>0.056117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>6578930</td>\n",
       "      <td>0.002110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RC transposons</td>\n",
       "      <td>314210</td>\n",
       "      <td>0.000101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           feature        Len  proportion\n",
       "0  DNA transposons   62872991    0.020169\n",
       "1             LINE  464215444    0.148917\n",
       "2              LTR  174930907    0.056117\n",
       "3   Low_complexity    6578930    0.002110\n",
       "4   RC transposons     314210    0.000101"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RM_contribution = RM.replace('-', 'Unknown')\n",
    "RM_contribution[['feature1', 'feature2']] = RM_contribution.feature.str.split(\"/\", expand = True)\n",
    "RM_contribution['Len'] = RM_contribution[\"thickEnd\"] - RM_contribution['thickStart']\n",
    "RM_contribution = RM_contribution[['feature1', 'Len']]\n",
    "RM_contribution = RM_contribution.groupby(['feature1'])['Len'].sum()\n",
    "RM_contribution = RM_contribution.to_frame().reset_index()\n",
    "RM_contribution = RM_contribution.replace('DNA', 'DNA transposons')\n",
    "RM_contribution = RM_contribution.replace('RC', 'RC transposons')\n",
    "RM_contribution = RM_contribution.replace('Retroposon', 'SVA')\n",
    "RM_contribution = RM_contribution.rename(columns={'feature1':'feature'})\n",
    "print(RM_contribution.feature.unique())\n",
    "proportion = []\n",
    "for i in range(0, len(RM_contribution)):\n",
    "    proportion.append(RM_contribution.iloc[i]['Len']/genome_len)\n",
    "RM_contribution['proportion'] = proportion\n",
    "RM_contribution.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9b8b051f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CDS' 'exon' 'start_codon' 'stop_codon']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>Len</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDS</td>\n",
       "      <td>56116925</td>\n",
       "      <td>0.018002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon</td>\n",
       "      <td>176677617</td>\n",
       "      <td>0.056677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>start_codon</td>\n",
       "      <td>66051</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stop_codon</td>\n",
       "      <td>66064</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       feature        Len  proportion\n",
       "0          CDS   56116925    0.018002\n",
       "1         exon  176677617    0.056677\n",
       "2  start_codon      66051    0.000021\n",
       "3   stop_codon      66064    0.000021"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#genes\n",
    "genes['Len'] = genes[\"thickEnd\"] - genes['thickStart']\n",
    "genes_contribution = genes.groupby(['feature'])['Len'].sum()\n",
    "genes_contribution = genes_contribution.to_frame().reset_index()\n",
    "print(genes_contribution.feature.unique())\n",
    "proportion = []\n",
    "for i in range(0, len(genes_contribution)):\n",
    "    proportion.append(genes_contribution.iloc[i]['Len']/genome_len)\n",
    "genes_contribution['proportion'] = proportion\n",
    "genes_contribution.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "17039b6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9779912494170018"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#introns\n",
    "intron['Len'] = intron[\"thickEnd\"] - intron['thickStart']\n",
    "intron_sum = intron['Len'].sum()\n",
    "intron_contribution = intron_sum/genome_len\n",
    "intron_contribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "035c7fea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.051793982902122705"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#enhancers\n",
    "enhancer['Len'] = enhancer[\"thickEnd\"] - enhancer['thickStart']\n",
    "enhancer_sum = enhancer['Len'].sum()\n",
    "enhancer_contribution = enhancer_sum/genome_len\n",
    "enhancer_contribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "4dbbbf41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.019874883365337815"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#promoters\n",
    "promoter['Len'] = promoter[\"thickEnd\"] - promoter['thickStart']\n",
    "promoter_sum = promoter['Len'].sum()\n",
    "promoter_contribution = promoter_sum/genome_len\n",
    "promoter_contribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "04a13233",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>Len</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDS</td>\n",
       "      <td>56116925</td>\n",
       "      <td>0.018002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon</td>\n",
       "      <td>176677617</td>\n",
       "      <td>0.056677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>start_codon</td>\n",
       "      <td>66051</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stop_codon</td>\n",
       "      <td>66064</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNA transposons</td>\n",
       "      <td>62872991</td>\n",
       "      <td>0.020169</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           feature        Len  proportion\n",
       "0              CDS   56116925    0.018002\n",
       "1             exon  176677617    0.056677\n",
       "2      start_codon      66051    0.000021\n",
       "3       stop_codon      66064    0.000021\n",
       "0  DNA transposons   62872991    0.020169"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = {'feature': ['intron', 'promoter', 'enhancer'], 'Len': [intron_sum, promoter_sum, enhancer_sum], 'proportion': [intron_contribution, promoter_contribution, enhancer_contribution]}\n",
    "result = pd.DataFrame(data=d)\n",
    "\n",
    "data = [genes_contribution, RM_contribution, result]\n",
    "contributions = pd.concat(data)\n",
    "contributions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a585ed45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>number_allMRpur</th>\n",
       "      <th>number_allMRpur_pc</th>\n",
       "      <th>number_MRpur_inAGsegm</th>\n",
       "      <th>number_MRpur_inAGsegm_pc</th>\n",
       "      <th>numb_allMR_plusFeatureInput</th>\n",
       "      <th>numb_MRsegm_plusFeatureInput</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STR</td>\n",
       "      <td>CDS</td>\n",
       "      <td>78</td>\n",
       "      <td>0.042549</td>\n",
       "      <td>17</td>\n",
       "      <td>0.025745</td>\n",
       "      <td>2.363609</td>\n",
       "      <td>1.430153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STR</td>\n",
       "      <td>DNA transposons</td>\n",
       "      <td>225</td>\n",
       "      <td>0.122739</td>\n",
       "      <td>23</td>\n",
       "      <td>0.034832</td>\n",
       "      <td>6.085457</td>\n",
       "      <td>1.726995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STR</td>\n",
       "      <td>LINE</td>\n",
       "      <td>1865</td>\n",
       "      <td>1.017369</td>\n",
       "      <td>297</td>\n",
       "      <td>0.449789</td>\n",
       "      <td>6.831783</td>\n",
       "      <td>3.020398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STR</td>\n",
       "      <td>LTR</td>\n",
       "      <td>758</td>\n",
       "      <td>0.413494</td>\n",
       "      <td>83</td>\n",
       "      <td>0.125699</td>\n",
       "      <td>7.368473</td>\n",
       "      <td>2.239953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STR</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>30376</td>\n",
       "      <td>16.570294</td>\n",
       "      <td>22046</td>\n",
       "      <td>33.387348</td>\n",
       "      <td>7851.454762</td>\n",
       "      <td>15819.831357</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  repeat_type          feature  number_allMRpur  number_allMRpur_pc  \\\n",
       "0         STR              CDS               78            0.042549   \n",
       "1         STR  DNA transposons              225            0.122739   \n",
       "2         STR             LINE             1865            1.017369   \n",
       "3         STR              LTR              758            0.413494   \n",
       "4         STR   Low_complexity            30376           16.570294   \n",
       "\n",
       "   number_MRpur_inAGsegm  number_MRpur_inAGsegm_pc  \\\n",
       "0                     17                  0.025745   \n",
       "1                     23                  0.034832   \n",
       "2                    297                  0.449789   \n",
       "3                     83                  0.125699   \n",
       "4                  22046                 33.387348   \n",
       "\n",
       "   numb_allMR_plusFeatureInput  numb_MRsegm_plusFeatureInput  \n",
       "0                     2.363609                      1.430153  \n",
       "1                     6.085457                      1.726995  \n",
       "2                     6.831783                      3.020398  \n",
       "3                     7.368473                      2.239953  \n",
       "4                  7851.454762                  15819.831357  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3 = df3[df3['feature']!='Unknown']\n",
    "numb_allMR_plusFeatureInput = []\n",
    "numb_MRsegm_plusFeatureInput = []\n",
    "for i in range(0, len(df3)):\n",
    "    feature = df3.iloc[i]['feature']\n",
    "    df = contributions[contributions[\"feature\"]==feature]\n",
    "    contribution = df.iloc[0]['proportion']\n",
    "    numb_allMR_plusFeatureInput.append((df3.iloc[i]['number_allMRpur_pc'])/contribution)\n",
    "    numb_MRsegm_plusFeatureInput.append((df3.iloc[i]['number_MRpur_inAGsegm_pc'])/contribution)\n",
    "df3['numb_allMR_plusFeatureInput'] = numb_allMR_plusFeatureInput\n",
    "df3['numb_MRsegm_plusFeatureInput'] = numb_MRsegm_plusFeatureInput\n",
    "#df.to_csv('num_MRrep_featuresContributions.csv', sep='\\t', index=False)\n",
    "df3.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0eee6627",
   "metadata": {},
   "source": [
    "### Transposons\n",
    "\n",
    "<a name='TE'></a>\n",
    "[Return to Table of Contents](#TOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9af32a30",
   "metadata": {},
   "outputs": [],
   "source": [
    "transposons = ['SINE', 'LTR', 'LINE', 'DNA', 'Retroposon', 'RC']\n",
    "\n",
    "MR_annotated = pd.read_csv('data/outputs/MRpur_annotated.csv', sep='\\t')\n",
    "MRsegm_annotated = pd.read_csv('data/outputs/MRpur_inAGsegm_annotated.csv', sep='\\t')\n",
    "\n",
    "df = MR_annotated.replace('-', 'Unknown')\n",
    "df['feature'] = df.feature.str.split('; ')\n",
    "df = df.explode(\"feature\")\n",
    "df[['feature1', 'feature2']] = df.feature.str.split(\"/\", expand = True)\n",
    "df = df.reset_index(drop=True)\n",
    "df = df.reset_index()\n",
    "MR_transposons = df[df['feature1'].isin(transposons)]\n",
    "\n",
    "df2 = MRsegm_annotated.replace('-', 'Unknown')\n",
    "df2['feature'] = df2.feature.str.split('; ')\n",
    "df2 = df2.explode(\"feature\")\n",
    "df2[['feature1', 'feature2']] = df2.feature.str.split(\"/\", expand = True)\n",
    "df2 = df2.reset_index(drop=True)\n",
    "df2 = df2.reset_index()\n",
    "MRsegm_transposons = df2[df2['feature1'].isin(transposons)]\n",
    "\n",
    "MR_transposons = MR_transposons.groupby(['lsh'])['index'].count()\n",
    "MR_transposons = MR_transposons.to_frame().reset_index()\n",
    "MR_transposons.to_csv('data/outputs/transposons_number.csv', sep='\\t', index=False)\n",
    "MRsegm_transposons = MRsegm_transposons.groupby(['lsh'])['index'].count()\n",
    "MRsegm_transposons = MRsegm_transposons.to_frame().reset_index()\n",
    "MRsegm_transposons.to_csv('data/outputs/transposons_segm_number.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "531cf576",
   "metadata": {},
   "source": [
    "### ISTAT\n",
    "\n",
    "<a name='istat'></a>\n",
    "[Return to Table of Contents](#TOC)\n",
    "- finding the most common features for checking overlaps significance  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "cd952446",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeat_type</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>name</th>\n",
       "      <th>number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>short_complex</td>\n",
       "      <td>SINE</td>\n",
       "      <td>Alu</td>\n",
       "      <td>AluSx</td>\n",
       "      <td>3514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>short_complex</td>\n",
       "      <td>SINE</td>\n",
       "      <td>Alu</td>\n",
       "      <td>AluJb</td>\n",
       "      <td>1988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>short_complex</td>\n",
       "      <td>SINE</td>\n",
       "      <td>Alu</td>\n",
       "      <td>AluY</td>\n",
       "      <td>1476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>short_complex</td>\n",
       "      <td>SINE</td>\n",
       "      <td>Alu</td>\n",
       "      <td>AluSx1</td>\n",
       "      <td>625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>short_complex</td>\n",
       "      <td>SINE</td>\n",
       "      <td>Alu</td>\n",
       "      <td>AluSz</td>\n",
       "      <td>527</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       repeat_type class family    name  number\n",
       "642  short_complex  SINE    Alu   AluSx    3514\n",
       "627  short_complex  SINE    Alu   AluJb    1988\n",
       "648  short_complex  SINE    Alu    AluY    1476\n",
       "643  short_complex  SINE    Alu  AluSx1     625\n",
       "646  short_complex  SINE    Alu   AluSz     527"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('MRpur_annotated.csv', sep='\\t')\n",
    "df['feature'] = df.feature.str.split('; ')\n",
    "df = df.explode(\"feature\")\n",
    "df['name'] = df.name.str.split('; ')\n",
    "df2 = df.explode(\"name\")\n",
    "df2[['feature1', 'feature2']] = df2.feature.str.split(\"/\", expand = True)\n",
    "df2 = df2.reset_index(drop=True)\n",
    "df2 = df2.reset_index()\n",
    "df2 = df2[~df2['feature1'].isin(['Simple_repeat', 'Low_complexity', 'DNA', 'srpRNA', 'scRNA', 'Satellite', '-', 'Unknown', 'RC'])]\n",
    "df2 = df2[df2['name']!='-']\n",
    "df3 = df2.groupby(['repeat_type', 'feature1', 'feature2', 'name'])['index'].count()\n",
    "df3 = df3.to_frame().reset_index()\n",
    "df3 = df3.rename(columns={\"index\": \"number\", 'feature1':'class', 'feature2':'family'})\n",
    "df3 = df3.sort_values(by=['repeat_type', 'number'], ascending=False)\n",
    "df3 = df3\n",
    "df3.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a3dddee",
   "metadata": {},
   "source": [
    "## Conservation\n",
    "\n",
    "<a name='conservation'></a>\n",
    "[Return to Table of Contents](#TOC)\n",
    "\n",
    "conservation score phyloP for multiple alignment by chromosomes http://hgdownload.soe.ucsc.edu/goldenPath/hg38/phyloP30way/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f787c334",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function for reading wigFix files\n",
    "flatten = lambda l: [item for sublist in l for item in sublist]\n",
    "def read_wig_fixed(file):\n",
    "    phylo = pd.read_csv(file, sep = ',', compression = 'gzip', header = None, dtype = str)\n",
    "    phylo_index_entries = phylo.loc[[entry.startswith('fixedStep') for entry in phylo[0].astype(str)]][0].str.split('start=', expand = True)[1].str.split(' ', expand = True)[0].astype(int)\n",
    "    phylo_index_end = list(phylo_index_entries.index[1:]) + [phylo.index.max()]\n",
    "    phylo_new_index = flatten([list(range(entry-1, entry+(end-start)-1)) for start, end, entry in zip(phylo_index_entries.index, phylo_index_end, phylo_index_entries)]) + list(phylo_index_entries[-1:])\n",
    "    phylo.index = phylo_new_index\n",
    "    phylo = pd.to_numeric(phylo[0], errors='coerce').dropna()\n",
    "    phylo = phylo.loc[~phylo.index.duplicated()].sort_index()\n",
    "    return phylo\n",
    "\n",
    "def main():\n",
    "    for Chr in chr_list_XY:\n",
    "        print(Chr, end=' ')\n",
    "        infile = 'conservation/'+Chr+'.phyloP30way.wigFix.gz'\n",
    "        outfile = 'conservation/'+Chr+'_phylo.txt'\n",
    "\n",
    "        df = read_wig_fixed(infile)\n",
    "        df.to_csv(outfile, sep='\\t')\n",
    "                \n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9199480f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MR is file with repeats where names of columns are chr, start of repeat, end of repeat\n",
    "MR = pd.read_csv('data/MR_hg38.bed', sep='\\t', names=['chr', 'start', 'end'])\n",
    "MR['length'] = MR['end'] - MR['start']\n",
    "score_list = []\n",
    "\n",
    "for Chr in chr_list_XY:\n",
    "    print(Chr, end=' ')\n",
    "    MR_chr = MR[MR['chr'] == Chr]\n",
    "    MR_chr = MR_chr.reset_index()\n",
    "    \n",
    "    infile = 'conservation/' + Chr + '_phylo.txt'\n",
    "    df = pd.read_csv(infile, sep='\\t', names = ['position', 'score'])\n",
    "    \n",
    "    for row in range(0, len(MR_chr)):\n",
    "        start = MR_chr.iloc[row]['start']\n",
    "        end = MR_chr.iloc[row]['end']\n",
    "        length = MR_chr.iloc[row]['length']\n",
    "        score_sum = 0\n",
    "        for n in range(start, end):\n",
    "            try:\n",
    "                score_sum += df.at[n, 'score']\n",
    "            except:\n",
    "                length -= 1\n",
    "                continue\n",
    "        if length == 0:\n",
    "            score_list.append(0)\n",
    "        else:\n",
    "            score_list.append(score_sum/length)\n",
    "\n",
    "MR['conservation_score'] = score_list\n",
    "MR.to_csv('data/conservation_score_MR.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "548a7b41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# average conservation\n",
    "mean_score = 0\n",
    "for Chr in chr_list_XY:\n",
    "    print(Chr, end=' ')\n",
    "    infile = '../new/conservation/' + Chr + '_phylo.txt'\n",
    "    df = pd.read_csv(infile, sep='\\t', names = ['position', 'score'])\n",
    "    \n",
    "    score_sum = df.sum()['score']\n",
    "    mean_score += score_sum/len(df)\n",
    "    \n",
    "mean_score = mean_score/len(chr_list_XY)\n",
    "        \n",
    "print(mean_score)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
